include/device.hpp

```diff
@@ -0,0 +1,574 @@
+#define FEATURE_PIOQUIET 0x00040000
+
+#define FEATURE_LOEJ 0x00080000
+#define FEATURE_LOEJ_TOGGLE 0x00100000
+#define FEATURE_LOCK 0x00200000
+#define FEATURE_LOCK_TOGGLE 0x00400000
+
+#define AS_ACTION_MODE 1
+#define AS_ACTION_ACT 2
+#define AS_ACTION_DEACT 3
+#define AS_ACTION_DEL 4
+#define AS_ACTION_CLEAR 5
+#define AS_ACTION_CRE 6
+#define AS_ACTION_MQCK 7
+
+#define AS_MODE_OFF 0
+#define AS_MODE_AUTO 1
+#define AS_MODE_ON 2
+#define AS_MODE_FORCED 3
+
+#define ASCRE_FULL 8
+#define ASCRE_REPLACE 16
+
+#define ASMQCK_ADV 32
+
+#define PIOQ_QUIET 0
+#define PIOQ_STD 1
+#define PIOQ_PERF 2
+
+#define QPX_FILE_SIGN "QPXD"
+
+#if 0
+// chunk. cnout = 1
+#define META_TEST_ERRC_SUMMARY 57
+#define META_TEST_JB_SUMMARY 58
+#define META_TEST_FT_SUMMARY 59
+#endif
+
+struct ASDB_item {
+  bool present, active;
+  QString type;
+  QString mid, speed, writes;
+};
+
+typedef QList<ASDB_item> ASDB;
+
+struct DevFeatures {
+  int supported;
+  int enabled;
+
+  int prec_spd;
+  float grec;
+  char vrec_cd_pwr;
+  int vrec_cd_str;
+  char vrec_dvd_pwr;
+  int vrec_dvd_str;
+
+  QString sr_pass;
+
+  int sm_cd_rd, sm_cd_wr, sm_dvd_rd;
+  bool sm_access; // 0 = slow, 1 = fast
+  int sm_trayl, sm_traye;
+  bool sm_nosave;
+
+  int psm_cd_rd, psm_cd_wr, psm_dvd_rd;
+  //			psm_dvd_wr;
+  bool psm_access; // 0 = slow, 1 = fast
+  int psm_trayl, psm_traye;
+
+  int tattoo_inner, tattoo_outer;
+  QString tattoo_file;
+
+  int as_action;
+  int as_act_mode;
+  int as_mode;
+  int as_idx;
+  int as_mqckspd;
+  QString as_mqckres;
+  int pioq_quiet;
+  bool pioq_nosave;
+};
+
+struct DI_Transfer {
+  quint64 lba;
+  float spdx;
+  uint32_t spdk;
+};
+
+template <typename T> struct Errc_ARRAY {
+  quint64 lba;
+  float spdx;
+  T err[8];
+};
+
+template <typename T> struct Errc_CD {
+  quint64 lba;
+  float spdx;
+  T bler;
+  T e11, e21, e31;
+  T e12, e22, e32;
+  T uncr;
+};
+
+template <typename T> struct Errc_DVD {
+  quint64 lba;
+  float spdx;
+  T res;
+  T pie, pi8, pif;
+  T poe, po8, pof;
+  T uncr;
+};
+
+template <typename T> struct Errc_BD {
+  quint64 lba;
+  float spdx;
+  T res0;
+  T ldc, res1, res2;
+  T bis, res3, res4;
+  T uncr;
+};
+
+template <typename T> union Errc {
+  Errc_ARRAY<T> raw;
+  Errc_CD<T> cd;
+  Errc_DVD<T> dvd;
+  Errc_BD<T> bd;
+};
+
+typedef Errc<int> DI_Errc;
+typedef Errc<int64_t> TOT_Errc;
+typedef Errc<float> AVG_Errc;
+
+void ErrcADD(TOT_Errc *tot, const DI_Errc &o);
+void ErrcMAX(DI_Errc *max, const DI_Errc &o);
+void CDErrcAVG(AVG_Errc *avg, TOT_Errc *tot, uint64_t blocks);
+void DVDErrcAVG(AVG_Errc *avg, TOT_Errc *tot, uint64_t blocks);
+void BDErrcAVG(AVG_Errc *avg, TOT_Errc *tot, uint64_t blocks);
+
+struct DI_JB {
+  quint64 lba;
+  float spdx;
+  float jitter;
+  float asymm;
+};
+
+struct MM_JB {
+  float jmin;
+  float jmax;
+  float bmin;
+  float bmax;
+};
+
+struct DI_FT {
+  quint64 lba;
+  float spdx;
+  quint32 fe;
+  quint32 te;
+};
+
+struct DI_TA {
+  // float T;
+  int idx;
+  int pit;
```

Comment:
Codacy found an issue: [struct member 'DI_TA::pit' is never used.](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)